### PR TITLE
Add domain unix socket supports

### PR DESCRIPTION
**Pull request by XxChang**
The unix domain socket is more efficient than TCP socket in local linux machine.

The PR add support to unix domain socket.

### CODE REVIEW

**Review by phil-opp — COMMENTED** [review]
Thanks a lot, looks very good overall!

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -129,6 +129,11 @@ jobs:
         if: runner.os == 'Linux'
         timeout-minutes: 30
         run: cargo run --example cmake-dataflow
+      - name: "Unix Domain Socket example"
+        if: runner.os == 'Linux'
+        run: |
+          sed -i '1i communication:\n  _unstable_local:\n    UnixDomain\n' examples/rust-dataflow/dataflow.yml
+          cargo run --example rust-dataflow
```
[the remark this answers is not in the document]
Comment (XxChang) [reply .github/workflows/ci.yml:136]: Good point

**Review by phil-opp — APPROVED** [review]
Thanks for the quick update!

**Comment by haixuanTao** [comment]
So, on my local test, I didn't see any difference in performance running dora-benchmark.

Wonder if this makes a difference on very specialized hardware.

**Comment by XxChang** [comment]
> So, on my local test, I didn't see any difference in performance running dora-benchmark.
> 
> Wonder if this makes a difference on very specialized hardware.

Thanks for your testing! That's my mistake. It is because that the `write` method is blocked by UnixStream. 

I will update it ASAP.

The following is UnixStream with `set_nonblocking(true)`
```
Latency:
size 0x0     : 632.856µs
size 0x8     : 633.242µs
size 0x40    : 843.875µs
size 0x200   : 796.246µs
size 0x800   : 1.177051ms
size 0x1000  : 603.893µs
size 0x4000  : 851.743µs
size 0xa000  : 489.578µs
size 0x64000 : 143.565254ms
size 0x3e8000: 5.95041138s
Throughput:
size 0x0     : 19565 messages per second
size 0x8     : 17443 messages per second
size 0x40    : 12432 messages per second
size 0x200   : 5754 messages per second
size 0x800   : 2374 messages per second
size 0x1000  : 4059 messages per second
size 0x4000  : 1105 messages per second
size 0xa000  : 543 messages per second
size 0x64000 : 56 messages per second
Input `latency` was closed
Input `throughput` was closed
size 0x3e8000: 7 messages per second
```

the following is TcpStream:
```
Latency:
size 0x0     : 952.865µs
size 0x8     : 974.297µs
size 0x40    : 402.752µs
size 0x200   : 657.945µs
size 0x800   : 1.223403ms
size 0x1000  : 414.758µs
size 0x4000  : 538.546µs
size 0xa000  : 499.841µs
size 0x64000 : 203.923098ms
size 0x3e8000: 7.378916741s
Throughput:
size 0x0     : 14439 messages per second
size 0x8     : 12762 messages per second
size 0x40    : 13734 messages per second
size 0x200   : 6339 messages per second
size 0x800   : 2075 messages per second
size 0x1000  : 4066 messages per second
size 0x4000  : 1333 messages per second
size 0xa000  : 360 messages per second
size 0x64000 : 39 messages per second
Input `latency` was closed
Input `throughput` was closed
size 0x3e8000: 6 messages per second
```

**Comment by XxChang** [comment]
> So, on my local test, I didn't see any difference in performance running dora-benchmark.
> 
> Wonder if this makes a difference on very specialized hardware.

I also did a test, got the same result. It's wired. In theory, it should be more efficient than Tcp socket. [More details](https://stackoverflow.com/questions/14973942/tcp-loopback-connection-vs-unix-domain-socket-performance).

the following is UnixStream with `set_nonblocking(true)`
```
Latency:
size 0x0     : 537.864µs
size 0x8     : 674.116µs
size 0x40    : 397.551µs
size 0x200   : 566.487µs
size 0x800   : 1.306532ms
size 0x1000  : 462.612µs
size 0x4000  : 461.783µs
size 0xa000  : 398.878µs
size 0x64000 : 148.14771ms
size 0x3e8000: 6.20969162s
Throughput:
size 0x0     : 16925 messages per second
size 0x8     : 15287 messages per second
size 0x40    : 14140 messages per second
size 0x200   : 5721 messages per second
size 0x800   : 1886 messages per second
size 0x1000  : 3967 messages per second
size 0x4000  : 1166 messages per second
size 0xa000  : 451 messages per second
size 0x64000 : 59 messages per second
Input `latency` was closed
Input `throughput` was closed
size 0x3e8000: 7 messages per second
```

the following is UnixStream
```
Latency:
size 0x0     : 335.269µs
size 0x8     : 341.425µs
size 0x40    : 339.099µs
size 0x200   : 541.176µs
size 0x800   : 1.13206ms
size 0x1000  : 373.098µs
size 0x4000  : 354.6µs
size 0xa000  : 381.095µs
size 0x64000 : 236.397237ms
size 0x3e8000: 6.359789191s
Throughput:
size 0x0     : 754 messages per second
size 0x8     : 19929 messages per second
size 0x40    : 16591 messages per second
size 0x200   : 5783 messages per second
size 0x800   : 2154 messages per second
size 0x1000  : 4186 messages per second
size 0x4000  : 960 messages per second
size 0xa000  : 473 messages per second
size 0x64000 : 55 messages per second
Input `latency` was closed
Input `throughput` was closed
size 0x3e8000: 6 messages per second
```

the following is TcpStream
```
Latency:
size 0x0     : 425.477µs
size 0x8     : 454.158µs
size 0x40    : 485.851µs
size 0x200   : 625.459µs
size 0x800   : 1.45471ms
size 0x1000  : 504.22µs
size 0x4000  : 505.09µs
size 0xa000  : 549.074µs
size 0x64000 : 214.217185ms
size 0x3e8000: 6.407898767s
Throughput:
size 0x0     : 20018 messages per second
size 0x8     : 18672 messages per second
size 0x40    : 18060 messages per second
size 0x200   : 6620 messages per second
size 0x800   : 2312 messages per second
size 0x1000  : 3859 messages per second
size 0x4000  : 1342 messages per second
size 0xa000  : 501 messages per second
size 0x64000 : 53 messages per second
Input `latency` was closed
Input `throughput` was closed
size 0x3e8000: 6 messages per second
```

**Comment by phil-opp** [comment]
The overhead of TCP headers etc is probably small enough to not matter too much in our case. One of the comments in the stackoverflow post you linked reported 6us latency for TCP and 2us latency for Unix domain sockets. We have latencies of at least 400us, so the 4us difference don't matter much.

**Pull request merged**